master/internal/task/all_gather.go

```diff
@@ -0,0 +1,169 @@
+	WatchAllGather struct {
+		WatcherID uuid.UUID
+		NumPeers  int
+		Data      *structpb.Struct
+	}
+	// UnwatchAllGather indicates the peer has disconnected.
+	UnwatchAllGather struct {
+		WatcherID uuid.UUID
+	}
+	// Indicates the all gather has timed out.
+	allGatherTimeout struct {
+		id uuid.UUID
+	}
+)
+
+// NewAllGather returns a new all gather component.
+func NewAllGather() *AllGather {
+	return &AllGather{
+		id:       uuid.New(),
+		watchers: map[uuid.UUID]chan<- AllGatherInfoOrError{},
+	}
+}
+
+// PreStart just steps up the rendezvous watcher.
+func (g *AllGather) PreStart(ctx *actor.Context) {
+	actors.NotifyAfter(ctx, DefaultAllGatherTimeout, allGatherTimeout{id: g.id})
+}
+
+// ReceiveMsg receives rendezvous-specific messages.
+func (g *AllGather) ReceiveMsg(ctx *actor.Context) (bool, error) {
+	switch msg := ctx.Message().(type) {
+	case WatchAllGather:
+		ctx.Respond(g.watch(msg.WatcherID, msg.NumPeers, msg.Data))
+	case UnwatchAllGather:
+		g.unwatch(msg.WatcherID)
+	case allGatherTimeout:
+		if err := g.checkTimeout(msg.id); err != nil {
+			return false, err
+		}
+	default:
+		return false, actor.ErrUnexpectedMessage(ctx)
+	}
+	return g.ready(), nil
+}
+
+func (g *AllGather) watch(id uuid.UUID, count int, data *structpb.Struct) AllGatherWatcher {
+	// Channel is size 1 since data info will only ever be sent once and we'd rather not block.
+	w := make(chan AllGatherInfoOrError, 1)
+	g.watchers[id] = w
+	g.data = append(g.data, data)
+	g.numPeers = ptrs.IntPtr(count)
+	g.lastPeerJoin = time.Now().UTC()
+	if g.ready() {
+		g.push()
+	}
+	return AllGatherWatcher{C: w}
+}
+
+func (g *AllGather) unwatch(id uuid.UUID) {
+	delete(g.watchers, id)
+}
+
+// ready returns true if and only if all the containers are reported to be started with the
+// ContainerStarted message and their sockets to be connected with the containerConnected
+// message. The two messages are not guaranteed to come in-order. During each run of the
+// trial, once all the containers are ready this function will return true afterward because this
+// function is used in deciding if the trial should be forcibly killed when terminating.
+func (g *AllGather) ready() bool {
+	if g == nil {
+		return false
+	}
+
+	if g.readyPassed {
+		return true
+	}
+
+	g.readyPassed = g.numPeers != nil && len(g.watchers) == *g.numPeers
+	return g.readyPassed
+}
+
+// push gathers up the external addresses for the exposed ports and sends them to all the
+// containers in the trial.
+func (g AllGather) push() bool {
+	if !g.ready() {
+		return false
+	}
+	for id, c := range g.watchers {
+		c <- AllGatherInfoOrError{Data: g.data}
+		close(c)
+		delete(g.watchers, id)
+	}
+	return true
+}
+
+// checkTimeout checks if the task should timeout waiting for rendezvous.
+func (g *AllGather) checkTimeout(id uuid.UUID) error {
+	if g == nil {
+		return nil
+	}
+
+	if g.id == id && time.Now().UTC().After(g.lastPeerJoin.Add(DefaultAllGatherTimeout)) {
```

Comment:
this is checking for timeout after the last peer, but `allGatherTimeout` is only triggered once in `PreStart`. should it be fired off periodically instead, or after every `.watch` call?

Reply:
my thought was just if an all gather takes longer than 10m from start to finish it prints the timeout, rather than adding any logic to restart the timer.

Reply:
then maybe this should just check if any peers have joined at all? because if anyone has joined then lastPeerJoin will always be less than DefaultAllGatherTimeout.

Reply:
oh i see what you mean, duh. i think i'll just rm lastPeerJoin and make it allGatherDeadline (time.Now() at start + timeout). sorry for confusion.